bibtasklets/bst_orcidsync.py

```diff
@@ -44,24 +44,63 @@
 from invenio.dbquery import run_sql
 from invenio.webuser import collect_user_info
 from invenio.bibtask import write_message
+from invenio.bibauthorid_webapi import get_hepnames
 
 CFG_REDIS_ORCID_SYNC_KEY = 'legacy_orcid_tokens'
 CFG_REDIS_ORCID_SYNC_KEY_COUNT = 'legacy_orcid_tokens_count'
 
 
 def orcid_data_to_sync():
-    """Returns an iterator of data to sync."""
+    """
+    Collect all available OAuth tokens for ORCID and their related data:
+    (orcid, token, email, name)
+    Orcid and token are mandatory data.
+    Email and name can be empty strings.
+    In order to collect the email and the name, the strategy is:
+    - if a uid is found, then get the email and name from the user table.
+    - otherwise, get the email and name from HepNames.
+
+    Note regarding account types.
+    There are 2 different types of account involved.
+    1. The Inspire user account that requires an email and name and assigns
+    a uid (in particular when users log in via arXiv). The BAI system personid
+    is separate and independent of the Inspire user accounts uid, but if a
+    connection can be made to a user account the uid is associated with the
+    personid in the aidPERSONIDDATA table.
+    2. HepNames which has personal data like name and email and is associated
+    with the BAI via MARC: 000996436 035__ $$9BAI$$aSimona.Murgia.1
+    There are some convenience methods to get info from HepNames:
+    from invenio.bibauthorid_webapi import get_hepnames
+    """
     rows = run_sql("SELECT personid, data, token FROM aidPERSONIDDATA NATURAL JOIN aidTOKEN WHERE tag='extid:ORCID';")
     for personid, orcid, token in rows:
+        orcid = orcid.decode('utf8')
+        token = token.decode('utf8')
+        email = ''
+        name = ''
+
         uid = run_sql("SELECT data FROM aidPERSONIDDATA WHERE personid=%s AND tag='uid'", (personid, ))
+
+        # If uid, then collect name and email from the user table.
         if uid:
             uid = int(uid[0][0])
             user_info = collect_user_info(uid)
+            email = user_info['email'].decode('utf8')
             if user_info.get('external_firstname') and user_info.get('external_familyname'):
                 name = u'%s %s' % (user_info['external_firstname'].decode('utf8'), user_info['external_familyname'].decode('utf8'))
             else:
                 name = user_info['nickname'].decode('utf8')
-            yield orcid.decode('utf8'), token.decode('utf8'), user_info['email'].decode('utf8'), name
+
+        # If not uid, then collect name and email from HepNames.
+        else:
+            hepnames = get_hepnames(personid)
+            try:
+                email = hepnames['record']['emails'][0].decode('utf8')
+                name = hepnames['record']['display_name'].decode('utf8')
```

Comment:
so when there is no `email`, the `name` will remain empty, too, even though it likely is available (HepNames should always have a `display_name`). is that intentional?

Reply:
Hmm, you are right. No, it is not intentional and I will improve this.